stat_flags.3
.\" -
.\" SPDX-License-Identifier: BSD-2-Clause
.\" Copyright (c) 1996 The NetBSD Foundation, Inc.  All rights reserved.
.\" This code is derived from software contributed to The NetBSD Foundation
.\" by Christos Zoulas.
.\"
.\" from NetBSD: stat_flags.3,v 1.8 2012/09/19 23:22:56 wiz Exp
.\"
.Dd September 20, 2012
.Dt STAT_FLAGS 3
.Os
.Sh NAME
.Nm string_to_flags ,
.Nm flags_to_string
.Nd Stat flags parsing and printing functions
.Sh LIBRARY
.Lb libutil
.Sh SYNOPSIS
.In util.h
.Ft char *
.Fn flags_to_string "u_long flags" "const char *def"
.Ft int
.Fn string_to_flags "char **stringp" "u_long *setp" "u_long *clrp"
.Sh DESCRIPTION
The
.Fn flags_to_string
and
.Fn string_to_flags
functions are used by
programs such as
.Xr ls 1 ,
.Xr mtree 8 ,
.Xr makefs 8 ,
etc., to parse and/or print the
.Dv st_flags field in the
.Xr stat 2
structure.
.Pp
They recognize the following flags:
.Bl -column -offset indent "uappnd " "SF_IMMUTABLE" "xxx"
.It Sy String Ta Sy Flag Ta Sy Description
.It Va arch Ta Dv SF_ARCHIVED Ta file is archived
.It Va nodump Ta Dv UF_NODUMP Ta do not dump file
.It Va opaque Ta Dv UF_OPAQUE Ta directory is opaque in union filesystems
.It Va sappnd Ta Dv SF_APPEND Ta writes to the file may only append
.It Va schg Ta Dv SF_IMMUTABLE Ta file cannot be changed; it is immutable
.It Va snap Ta Dv SF_SNAPSHOT Ta file is a snapshot inode
.It Va uappnd Ta Dv UF_APPEND Ta writes to the file may only append
.It Va uchg Ta Dv UF_IMMUTABLE Ta file cannot be changed; it is immutable
.El
.Pp
The
.Dv SF_APPEND
and
.Dv SF_IMMUTABLE
flags are for the superuser only, whereas
.Dv UF_APPEND
and
.Dv UF_IMMUTABLE
are for the user only.
.Pp
The
.Fn flags_to_string
function converts the bits set in the
.Fa flags
argument to a comma-separated string and returns it.
If no flags are set, then the
.Fa def
string is returned.
The returned string is allocated via
.Xr malloc 3
and it is the responsibility of the caller to
.Xr free 3
it.
.Pp
The
.Fn string_to_flags
function takes a
.Fa stringp
of space, comma, or tab separated flag names
and places their bit value on the
.Fa setp
argument.
If the flag name is prefixed by:
.Dq no ,
then the bit value is placed on the
.Fa clrp
argument.
.Sh RETURN VALUES
.Fn flags_to_string
returns the symbolic representation of flags, the default string, or
.Dv NULL
if allocation failed.
.Pp
.Fn string_to_flags
returns
.Dv 0
on success and
.Dv 1
if it fails to parse the string, setting
.Fa stringp
to point to the first string that it failed to parse.
.Sh SEE ALSO
.Xr chflags 2 ,
.Xr stat 2


HISTORY.rst
:SPDX-License-Identifier: BSD-2-Clause

This repository is very weird:
in an attempt to preserve as much of
the development history as possible,
I've synthesised commits from several sources,
including
`the CSRG archives <https://github.com/csrg/csrg>`_,
`FreeBSD <https://www.freebsd.org/>`_,
`NetBSD <https://www.netbsd.org/>`_,
`OpenBSD <https://www.openbsd.org/>`_,
`pkgsrc <https://www.netbsd.org/>`_,
`libarchive <https://github.com/libarchive/libarchive/>`_,
as well as porting and packaging efforts from
`Debian <https://github.com/jgoerzen/mtree-netbsd>`_ and
`OpenSUSE <https://github.com/archiecobbs/nmtree>`_.

Commit messages include annotations, beginning ``;;``,
with information about the commit itself.

These provenance markers may include:

``;; CSRG-SVN: rNNNNNN``:
   :Browse: <https://SVNweb.FreeBSD.org/csrg>

   Commit derived from CSRG data preserved by Kirk McKusick
   and converted from SCCS to Subversion by John Baldwin.
   Git conversions exist with varying levels of completeness;
   for example,
   <https://github.com/jonathangray/csrg>, or
   <https://github.com/weiss/original-bsd> —
   these were useful for confirming human names.

``;; FreeBSD-Git: XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX``:
   :Browse: <https://cgit.FreeBSD.org/src/>
   :Browse: <https://github.com/freebsd/freebsd-src>

   Commit derived from FreeBSD at this Git revision.

``;; FreeBSD-SVN: rNNNNNN``:
   :Browse: <https://SVNweb.FreeBSD.org/base>

   Commit derived from FreeBSD at this Subversion revision.
   (Commits were synthesised from Git, not Subversion.
   CVS and Subversion keywords have varying levels of expansion.)

``;; NetBSD-CVS: PATH@rREV``:
   :Browse: <http://CVSweb.netbsd.org/bsdweb.cgi/src/>

   Commit derived from the NetBSD CVS repository at this point.
   (Commits were synthesised from Mercurial, not CVS,
   though I've checked the CVS artefacts for correctness.)

``;; NetBSD-Mercurial: rNNNNNN``:
   :Browse: <https://anonHg.NetBSD.org/src/>

   Commit derived from the NetBSD Mercurial repository at this revision.

``;; pkgsrc-CVS: PATH@rREV``:
   :Browse: <http://CVSweb.netbsd.org/bsdweb.cgi/pkgsrc/>

   Commit derived from the pkgsrc CVS repository at this point.
   (Commits were synthesised from Mercurial, not CVS,
   though I've checked the CVS artefacts for correctness.)

``;; pkgsrc-Mercurial: rNNNNNN``:
   :Browse: <https://anonHg.NetBSD.org/pkgsrc/>

   Commit derived from the pkgsrc Mercurial repository at this revision.

``;; OpenBSD-CVS: PATH@rREV``:
   :Browse: <https://CVSweb.OpenBSD.org/src/>

   Commit derived from the OpenBSD CVS repository at this point:
   these commits were synthesised from CVS artefacts.

``;; libarchive-Git: XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX``:
   :Browse: <https://github.com/libarchive/libarchive>

   Commit derived from the libarchive Git repository at this revision:
   given how few revisions there are here,
   I could export individual objects and
   generate the fast-import hints by hand.

``;; archiecobbs-Git: XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX``:
   :Browse: <https://github.com/archiecobbs/nmtree>

   Commit derived from Archie Cobbs' Git repository at this revision.

``;; jgoerzen-Git: XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX``:
   :Browse: <https://github.com/jgoerzen/mtree-netbsd>

   Commit derived from John Goerzen's Git repository at this revision.
   (That repository includes a Git import of the pkgsrc CVS history,
   which I've grafted by hand.)

Some version control systems — notably SCCS and CVS —
are bad at dealing with renaming files.
I've used context from around the commit
to reconstruct what filenames should be.
Notable manifestations of this are:

- ``create.c`` was originally called ``cwalk.c``.


stat_flags.c
/*-
 * SPDX-License-Identifier: BSD-3-Clause
 * Copyright (c) 1993 The Regents of the University of California
 *	All rights reserved.
 */

/*	from CSRG:   @(#)stat_flags.c	8.2 (Berkeley) 7/28/94 */
/*	from NetBSD: stat_flags.c,v 1.3 2022/04/19 20:32:17 rillig Exp */
/*	from pkgsrc: stat_flags.c,v 1.5 2011/07/27 15:31:00 seb Exp */

#if HAVE_CONFIG_H
#include "config.h"
#endif
#if 0
#if HAVE_NBTOOL_CONFIG_H
#include "nbtool_config.h"
#else
#define HAVE_STRUCT_STAT_ST_FLAGS 1
#endif
#endif
#if HAVE_NBCOMPAT_H
#include <nbcompat.h>
#endif

#if HAVE_SYS_CDEFS_H
#include <sys/cdefs.h>
#endif
#if !defined(lint)
#if 0
static char sccsid[] = "@(#)stat_flags.c	8.2 (Berkeley) 7/28/94";
#else
__RCSID("$NetBSD: stat_flags.c,v 1.5 2011/07/27 15:31:00 seb Exp $");
#endif
#endif /* not lint */

#if HAVE_SYS_TYPES_H
#include <sys/types.h>
#endif
#if HAVE_SYS_STAT_H
#include <sys/stat.h>
#endif

#if HAVE_FTS_H
#include <fts.h>
#endif
#if HAVE_STDDEF_H
#include <stddef.h>
#endif
#if HAVE_STRING_H
#include <string.h>
#endif

#include "stat_flags.h"

#define	SAPPEND(s) do {							\
	if (prefix != NULL)						\
		(void)strlcat(string, prefix, sizeof(string));		\
	(void)strlcat(string, s, sizeof(string));			\
	prefix = ",";							\
} while (/* CONSTCOND */ 0)

/*
 * flags_to_string --
 *	Convert stat flags to a comma-separated string.  If no flags
 *	are set, return the default string.
 */
char *
flags_to_string(u_long flags, const char *def)
{
	static char string[128];
	const char *prefix;

	string[0] = '\0';
	prefix = NULL;
#if HAVE_STRUCT_STAT_ST_FLAGS
	if (flags & UF_APPEND)
		SAPPEND("uappnd");
	if (flags & UF_IMMUTABLE)
		SAPPEND("uchg");
	if (flags & UF_NODUMP)
		SAPPEND("nodump");
	if (flags & UF_OPAQUE)
		SAPPEND("opaque");
	if (flags & SF_APPEND)
		SAPPEND("sappnd");
	if (flags & SF_ARCHIVED)
		SAPPEND("arch");
	if (flags & SF_IMMUTABLE)
		SAPPEND("schg");
#ifdef SF_SNAPSHOT
	if (flags & SF_SNAPSHOT)
		SAPPEND("snap");
#endif
#endif
	if (prefix == NULL)
		strlcpy(string, def, sizeof(string));
	return (string);
}

#define	TEST(a, b, f) {							\
	if (!strcmp(a, b)) {						\
		if (clear) {						\
			if (clrp)					\
				*clrp |= (f);				\
			if (setp)					\
				*setp &= ~(f);				\
		} else {						\
			if (setp)					\
				*setp |= (f);				\
			if (clrp)					\
				*clrp &= ~(f);				\
		}							\
		break;							\
	}								\
}

/*
 * string_to_flags --
 *	Take string of arguments and return stat flags.  Return 0 on
 *	success, 1 on failure.  On failure, stringp is set to point
 *	to the offending token.
 */
int
string_to_flags(char **stringp, u_long *setp, u_long *clrp)
{
	int clear;
	char *string, *p;

	if (setp)
		*setp = 0;
	if (clrp)
		*clrp = 0;

#if HAVE_STRUCT_STAT_ST_FLAGS
	string = *stringp;
	while ((p = strsep(&string, "\t ,")) != NULL) {
		clear = 0;
		*stringp = p;
		if (*p == '\0')
			continue;
		if (p[0] == 'n' && p[1] == 'o') {
			clear = 1;
			p += 2;
		}
		switch (p[0]) {
		case 'a':
			TEST(p, "arch", SF_ARCHIVED);
			TEST(p, "archived", SF_ARCHIVED);
			return (1);
		case 'd':
			clear = !clear;
			TEST(p, "dump", UF_NODUMP);
			return (1);
		case 'n':
				/*
				 * Support `nonodump'. Note that
				 * the state of clear is not changed.
				 */
			TEST(p, "nodump", UF_NODUMP);
			return (1);
		case 'o':
			TEST(p, "opaque", UF_OPAQUE);
			return (1);
		case 's':
			TEST(p, "sappnd", SF_APPEND);
			TEST(p, "sappend", SF_APPEND);
			TEST(p, "schg", SF_IMMUTABLE);
			TEST(p, "schange", SF_IMMUTABLE);
			TEST(p, "simmutable", SF_IMMUTABLE);
			return (1);
		case 'u':
			TEST(p, "uappnd", UF_APPEND);
			TEST(p, "uappend", UF_APPEND);
			TEST(p, "uchg", UF_IMMUTABLE);
			TEST(p, "uchange", UF_IMMUTABLE);
			TEST(p, "uimmutable", UF_IMMUTABLE);
			return (1);
		default:
			return (1);
		}
	}
#endif

	return (0);
}
